pysrc/utils/sampling_methods.py

```diff
@@ -1,29 +1,90 @@
 import math
 import random
+import pandas as pd
 from typing import Callable
 
 
 def random_sample(data: list[list[str]],
                   num_samples: int,
                   num_entries: int) -> list[str]:
-    return random.sample(data[0], k=num_samples)
+    tmp = pd.DataFrame(data[0])
+    tmp = tmp.replace(r'^s*$', float('NaN'), regex=True)
+    #tmp = tmp.dropna(inplace=True)
+
+    return tmp.sample(n=num_samples)
 
 
 def first_sample(data: list[list[str]],
                  num_samples: int,
                  num_entries: int) -> list[str]:
-    return data[0][:num_samples]
+    tmp = pd.DataFrame(data[0])
+
+    return tmp.iloc[:num_samples]
+
+#TODO Alle funktionen auf DataFrames umstellen sortieren und danach Empty Strings löschen führt zu mglw leereb Samples
+def smallest_value_sample(data: list[list[str]],
+                 num_samples: int,
+                 num_entries: int) -> list[str]:
+    tmp = pd.DataFrame(data[0])
+    tmp = tmp.replace(r'^s*$', float('NaN'), regex=True)
+    tmp.dropna(inplace=True)
+    for col in tmp.columns:
+        out = tmp.sort_values(by=col)
+
+    return out.iloc[:num_samples]
+
+def biggest_value_sample(data: list[list[str]],
+                 num_samples: int,
+                 num_entries: int) -> list[str]:
+    tmp = pd.DataFrame(data[0])
+    tmp = tmp.replace(r'^s*$', float('NaN'), regex=True)
+    tmp.dropna(inplace=True)
+    for col in tmp.columns:
+        out = tmp.sort_values(by=col, ascending=False)
+
+    return out.iloc[:num_samples]
+
+def longest_value_sample(data: list[list[str]],
+                 num_samples: int,
+                 num_entries: int) -> list[str]:
+    tmp = pd.DataFrame(data[0])
+    tmp = tmp.replace(r'^s*$', float('NaN'), regex=True)
+    tmp.dropna(inplace=True)
+    tmp.index = tmp[0].str.len()
+    df = tmp.sort_index(ascending=False).reset_index(drop=True)
+
+    return df.iloc[:num_samples]
+
+def shortest_value_sample(data: list[list[str]],
+                 num_samples: int,
+                 num_entries: int) -> list[str]:
+    tmp = pd.DataFrame(data[0])
+    tmp = tmp.replace(r'^s*$', float('NaN'), regex=True)
+    tmp.dropna(inplace=True)
+    tmp.index = tmp[0].str.len()
+    df = tmp.sort_index(ascending=True).reset_index(drop=True)
+
+    return df.iloc[:num_samples]
+
+#TODO rework that
+def all_distinct_sample(data: list[list[str]],
+                 num_samples: int,
+                 num_entries: int) -> list[str]:
+    tmp = pd.DataFrame(data[0])
+    df = tmp.loc[tmp.astype(str).drop_duplicates().index]
+
 
+    return df.iloc[:num_samples]
 
+#TODO Return Type is messed up
```

Comment:
Yeah, you have to completely update the types. But only do this when it's decided which types are actually going to be used.

Reply:
Not sure if it's worth to create a helper method for the two lines of code. I can add it for sure if you think its the better way.

Reply:
Ah, I think you misunderstood me. What I was trying to say was that the typings are (probably) incorrect now. You've changed the calling of the sampling methods in `sampling.py` to not use lists anymore (so like you've removed a layer of lists). The return type of the `iloc` calls are `Any` so I don't have any clue what they return. Apparently, they return something else now because you're using `iloc` on the return value of the sampling functions while they still claim to return `list[str]`.

Reply:
oh sorry no I wasn't entirely decided on the change for the types so far. My comment was more about the two lines I use to remove empty lines from the input. Surely I can move it to a helper function but not sure if it's worth to do so. It wouldn'T add much it might be a bit more readable that way.